packages/flutter/lib/src/material/page.dart

```diff
@@ -116,8 +118,76 @@ mixin MaterialRouteTransitionMixin<T> on PageRoute<T> {
 
   @override
   Widget buildTransitions(BuildContext context, Animation<double> animation, Animation<double> secondaryAnimation, Widget child) {
-    final PageTransitionsTheme theme = Theme.of(context).pageTransitionsTheme;
-    return theme.buildTransitions<T>(this, context, animation, secondaryAnimation, child);
+    final ThemeData themeData = Theme.of(context);
+    return ValueListenableBuilder<bool>(
+        valueListenable: navigator!.userGestureInProgressNotifier,
+        builder: (BuildContext context, bool useGestureInProgress, Widget? _) {
+          final bool usePrevTargetPlatform;
+          if (useGestureInProgress) {
+            // The platform should be kept unchanged during an user gesture.
+            usePrevTargetPlatform = _prevTargetPlatform != null && _prevTargetPlatform != themeData.platform;
+          } else {
+            _prevTargetPlatform = themeData.platform;
+            usePrevTargetPlatform = false;
+          }
+          return _PlatformOfBuilder(
+            platform: usePrevTargetPlatform ? _prevTargetPlatform : null,
+            builder: (BuildContext context, Widget? child) {
+              assert(child != null);
+              return themeData.pageTransitionsTheme.buildTransitions<T>(this, context, animation, secondaryAnimation, child!);
```

Comment:
I think rather than do this hack, I would create a new API in pageTransitionsTheme to buildTransition based on platform. so that you don't need to override the theme using _PlatformOfBuilder

Reply:
Ok, I removed `_PlatfrmOfBuilder` and added the `platform` parameter to `buildTransitions`.